Emurgo/project-icarus-importer
# WIP: #11 Slot and Ordinal

## Description

1. When storing successful txs - we also need to store the slot of the block they are included into, and their ordinal number inside that block.

2. When storing failed txs - we also need to assign it the latest known successful slot, so that failed tx can be ordered along the history of successful txs.

## Linked issue

Closing #11 

## Type of change
- [~] 🐞 Bug fix (non-breaking change which fixes an issue)
- [~] 🛠 New feature (non-breaking change which adds functionality)
- [~] 🏭 Refactoring that does not change existing functionality but does improve things like code readability, structure etc
- [~] 🔨 New or improved tests for existing code
- [~] ⛑ git-flow chore (backport, hotfix, etc)

## Developer checklist
- [~] If my code deals with exceptions, it follows the [guidelines](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/exceptions.md).
- [~] I have updated any documentation accordingly, if needed. Documentation changes can be reflected in opening a PR on [cardanodocs.com](https://github.com/input-output-hk/cardanodocs.com), amending the inline [Haddock](https://www.haskell.org/haddock/) comments, any relevant README file or one of the document listed in the [docs](https://github.com/input-output-hk/cardano-sl/tree/develop/docs) directory.

## Testing checklist
- [~] I have added tests to cover my changes.

## QA Steps
1. Drop existing database
2. Init new database from new SQL script (new columns added)
3. Run importer sync
4. Check importer can successfully store data into DB (no fails)
5. Check in DB that successful txs now contain epoch and slot numbers
6. Check that txs hash correct epoch and slot numbers (may be validated by block hash)
7. Check that successful txs now contain ordinal number
8. Check that txs contain right ordinal number (maybe checked by searching for multiple txs on the same Input/Output inside the same block)
9. Check that failed txs now contain epoch and slot number, but contain NO block hash and ordinal
10. Check that failed tx slot corresponds to closest previous successful tx slot
